### ISSUE TITLE
Round maps NaN to ±2^max_exp (Python/JS keep NaN)

### ISSUE THREAD

**Issue by SamirMoustafa** [SamirMoustafa/torch-floating-point#26]
## Problem

`Round` on a NaN input emits a large finite value instead of NaN. Both kernels take `floor(log2(|x|))` with no NaN guard, so `log2(NaN)` becomes NaN, `fmin`/`fmax` push the exponent to `max_exp`, and the output is `±2^max_exp`.

The Python clamp in `StraightThroughEstimator.forward` (`floating_point/round.py`) propagates NaN into `cpp_round`. There is no NaN test.

## Repro

```python
import math, torch
from floating_point import FloatingPoint, Round

fp16 = FloatingPoint(1, 5, 10, 15, 16)
y = Round(fp16)(torch.tensor([math.nan], dtype=torch.float32))
# observed: 32768.0  (and -NaN → -32768.0)
```

Same path in `floating_point/float_round.cpp` (CPU loop) and `floating_point/float_round_cuda.cu` (`float_round_one`).

## Expected

NaN in → NaN out, matching:

- `round_e0_scalar` / `round_scalar` in `scripts/gen_fp_codec_goldens.py` (`if math.isnan(x): return nan`)
- `round_scalar` in `docs/javascripts/fp-codec.js` (`Number.isNaN(x) → NaN`)
- decode: NaN codes already map to NaN (`floating_point/data_types.py`)

A NaN weight or activation in QAT must not become a silent finite spike.

## Proposal

Propagate NaN in both kernels (`if (x_val != x_val) return x_val;` / equivalent) before `log2`. Keep the signed-NaN bit if cheap (`copysign`), but do not require a distinction between quiet and signaling NaN.

## Acceptance

- `Round(fp)(NaN)` is NaN on CPU (and CUDA when that device is present) for a layout with a finite max (e.g. fp16 and E4M3-FN).
- `-NaN` does not become a finite negative value.
- Tests pin this on both devices when CUDA is available; CPU-only CI is enough for the CPU path.
- Python and JS scalar rounders stay NaN-preserving (already true).

## Out of scope

Inf (separate from NaN). Decode of NaN *codes* (already correct). GPU CI runner. Changing the clipped-STE mask for NaN inputs beyond “not a finite interior point.”
